### PR TITLE
feat(tsc): add tsc integration with opentelemetry sdk

### PR DESCRIPTION
**Pull request by Qinyouzeng**
integrate opentelemetry sdk to collecter logging,tracing and metrcis
 - tracing supports: aspnetcore , entityframework and http 
 - metrcis supports: runtime , aspnetcore and http

### CODE REVIEW

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/README.zh-CN.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+var ops = new OpenTelemetryInstrumentationOptions();
+//api exclude swagger and healthy request
+ops.AspNetCoreInstrumentationOptions.AppendDefaultFilter(ops);
+//blazor exclude blazor resources request
+//ops.AspNetCoreInstrumentationOptions.AppendBlazorFilter(ops);
+ops.BuildTraceCallback = builder => {
+    builder.SetResourceBuilder(resources);
+    builder.AddOtlpExporter();
+};
+builder.Services.AddMasaTracing(ops);
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/README.zh-CN.md:43]: Use option pattern.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/StreamExtenistions.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+namespace System.IO;
+
+public static class StreamExtenistions
+{
+    private static readonly Encoding _defaultEncoding = Encoding.UTF8;
+
+    public static async Task<string> ReadAsStringAsync(this Stream stream, Encoding? encoding = null)
+    {
+        if (stream == null)
+            return null;
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/StreamExtenistions.cs:13]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/StreamExtenistions.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            }            
+        } while (true);
+
+        if (data.Count > 0)
+        {
+            stream.Seek(start, SeekOrigin.Begin);
+            return (encoding ?? _defaultEncoding).GetString(data.ToArray());
+        }
+
+        return null;
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/StreamExtenistions.cs:46]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace System.Diagnostics;
+
+public static class ActivityExtension
+{
+    public static async Task<Activity> AddMasaSupplement(this Activity activity, HttpRequest httpRequest)
+    {
+        if (activity is null) return activity;
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs:10]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        activity.SetTag(OpenTelemetryAttributeName.Http.Flavor, httpRequest.Protocol);
+        activity.SetTag(OpenTelemetryAttributeName.Http.Scheme, httpRequest.Scheme);
+        activity.SetTag(OpenTelemetryAttributeName.Http.ClientIP, httpRequest.HttpContext?.Connection.RemoteIpAddress);
+        activity.SetTag(OpenTelemetryAttributeName.Http.RequestContentLength, httpRequest.ContentLength);
+        activity.SetTag(OpenTelemetryAttributeName.Http.RequestContentType, httpRequest.ContentType);
+        if (httpRequest.Body != null)
+        {
+            if (!httpRequest.Body.CanSeek)
+                httpRequest.EnableBuffering();
+            activity.SetTag(OpenTelemetryAttributeName.Http.RequestContentBody, await httpRequest.HttpContext.Request.Body.ReadAsStringAsync(GetHttpRequestEncoding(httpRequest)));
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs:21]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            var contentType = httpRequest.HttpContext.Request.ContentType;
+            if (!string.IsNullOrEmpty(contentType))
+            {
+                var attr = MediaTypeHeaderValue.Parse(contentType);
+                if (attr != null && !string.IsNullOrEmpty(attr.CharSet))
+                    return Encoding.GetEncoding(attr.CharSet);
+            }
+        }
+
+        return null;
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs:41]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public static Activity AddMasaSupplement(this Activity activity, HttpResponse httpResponse)
+    {
+        if (activity is null) return activity;
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs:46]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        activity.SetTag(OpenTelemetryAttributeName.Http.ResponseContentLength, httpResponse.ContentLength);
+        activity.SetTag(OpenTelemetryAttributeName.Http.ResponseContentType, httpResponse.ContentType);
+        activity.SetTag(OpenTelemetryAttributeName.Host.Name, Dns.GetHostName());
+
+        if ((httpResponse.HttpContext.User?.Claims.Count() ?? 0) > 0)
+        {
+            activity.AddTag(OpenTelemetryAttributeName.EndUser.Id, httpResponse.HttpContext.User.FindFirst("sub")?.Value ?? string.Empty);
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/ActivityExtensions.cs:54]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    /// </summary>
+    /// <param name="options"></param>
+    /// <param name="openTelemetryInstrumentationOptions"></param>
+    public static void AppendDefaultFilter(this Action<AspNetCoreInstrumentationOptions> options, OpenTelemetryInstrumentationOptions openTelemetryInstrumentationOptions)
+    {
+        options += opt =>
+        {
+            opt.Filter = httpContext =>
+            {
+                return !_lstDefaultFilterIgnorePrefix.Any(prefix => httpContext.Request.Path.Value.StartsWith(prefix));
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs:43]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    /// </summary>
+    /// <param name="options"></param>
+    /// <param name="openTelemetryInstrumentationOptions"></param>
+    public static void AppendBlazorFilter(this Action<AspNetCoreInstrumentationOptions> options, OpenTelemetryInstrumentationOptions openTelemetryInstrumentationOptions)
+    {
+        options += opt =>
+        {
+            opt.Filter = httpContext =>
+            {
+                if (_lstBlazorFilterIgnorePrefix.Any(prefix => httpContext.Request.Path.Value.StartsWith(prefix)))
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs:61]: Warning: Possible null reference return.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                    return false;
+
+                if (_lstBlazorFilterIgnoreSuffix.Any(suffix => httpContext.Request.Path.Value.EndsWith(suffix)))
```
Comment (doddgu) [inline src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Tracing/AspNetCoreInstrumentationOptionsExtensions.cs:64]: Warning: Possible null reference return.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/MasaObservableOptions.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public string ProjectName { get; set; }
+
+    public string ServiceName { get; set; }
+
+    public string ServiceNameSpace { get; set; }
+
+    public string ServiceInstanceId { get; set; }
+
+    public string ServerVersion { get; set; }
```
[the remark this answers is not in the document]
Comment (Qinyouzeng) [reply src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/MasaObservableOptions.cs:16]: It's service version.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/OpenTelemetryAttributeName.cs [hunk 13]
```diff
@@ -0,0 +1,146 @@
+// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Contrib.Observability;
```
[the remark this answers is not in the document]
Comment (Qinyouzeng) [reply src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/OpenTelemetryAttributeName.cs:4]: Change to Masa.Contrib.BasicAbility.Tsc.

File: src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Metrics/MasaServiceExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        services.AddOpenTelemetryMetrics(builder =>
+           {
+               if (builderConfiger != null)
+                   builderConfiger.Invoke(builder);
+
+               builder
+               .AddRuntimeMetrics()
+               .AddAspNetCoreInstrumentation()
+               .AddHttpClientInstrumentation();
```
[the remark this answers is not in the document]
Comment (Qinyouzeng) [reply src/BasicAbility/Masa.Contrib.BasicAbility.Tsc/Metrics/MasaServiceExtensions.cs:18]: Can use param `configure` type is `Action<MeterProviderBuilder>` , like this:
```
services.AddMasaMetrics(builder =>
{
    // add more instrumentation
    builder.AddCustomMetrics();
    builder.AddConsoleExporter();
}); 

```

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**